src/main/java/com/mixpanel/android/viewcrawler/ViewVisitor.java

```diff
@@ -152,19 +190,120 @@ public void accumulate(View found) {
                 LayoutRule originalValue = new LayoutRule(newVerb, currentRules[newVerb]);
                 mOriginalValues.put(found, originalValue);
             }
-            setLayout(found, newVerb, newAnchorId);
+
+            try {
+                setLayout(found, newVerb, newAnchorId);
+            } catch (CantVisitException e) {
+                cleanup();
+                mOnEditErrorListener.sendErrorMessage(e);
+            }
         }
 
-        private void setLayout(View target, int verb, int anchorId) {
+        private void setLayout(View target, int verb, int anchorId) throws CantVisitException {
             RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams)target.getLayoutParams();
             params.addRule(verb, anchorId);
+
+            final Set<Integer> rules;
+            if (mHorizontalRules.contains(verb)) {
+                rules = mHorizontalRules;
+            } else if (mVerticalRules.contains(verb)) {
+                rules = mVerticalRules;
+            } else {
+                rules = null;
+            }
+
+            if (rules != null && !verifyLayout(target, rules)) {
+                throw new CantVisitException("Circular dependency detected!", "circular_dependency", mName);
+            }
+
             target.setLayoutParams(params);
         }
 
+        private boolean verifyLayout(View target, Set<Integer> rules) {
+            ViewGroup parent = (ViewGroup) target.getParent();
+            SparseArray<View> idToChild = new SparseArray<View>();
+
+            int count = parent.getChildCount();
+            for (int i = 0; i < count; i++) {
```

Comment:
Can there be circular dependencies on views that aren't siblings? That is, can a view refer to another view outside of it's parent when aligning?